refnx/reduce/reduce.py

```diff
@@ -558,6 +559,439 @@ def _reduce_single_angle(self, scale=1):
+            Reduction options to apply to every spin channel being reduced.
+            This will override any individually defined reduction options
+            for each spin channel
+        """
+        # get a default set of reduction options
+        options = ReductionOptions()
+        options.update(reduction_options)
+
+        # set up the wavelength bins
+        if options["wavelength_bins"] is None:
+            wb = calculate_wavelength_bins(
+                options["lo_wavelength"],
+                options["hi_wavelength"],
+                options["rebin_percent"],
+            )
+            options["wavelength_bins"] = wb
+
+        reduced_successfully = []
+
+        # go through each spin channel and reduce it
+        for sc in ["dd", "du", "ud", "uu"]:
+            # first get the correct reduction options
+            rdo = spin_set_reflect.sc_opts[sc]
+            if rdo is None:
+                rdo = options
+            else:
+                # overwrite properties that need to be common
+                rdo["wavelength_bins"] = options["wavelength_bins"]
+
+            db = self.reducers[sc]
+            rb = getattr(spin_set_reflect, sc)
+            if rb is not None:
+                db.reduce(rb, save=False, **rdo)
+                getattr(self.spin_set_direct, sc).process(**rdo)
+                reduced_successfully.append(sc)
+            else:
+                # no reflected beam for a spin channel
+                continue
+
+        # by this point an unpolarised reduction has been done. But we need to
+        # correct the spectra for PNR. The following spectra (N, T) should be
+        # overwritten:
+        # TODO
+        # self.reducers["dd"].reflected_beam.m_spec
+        # self.reducers["dd"].reflected_beam.m_spec_sd
+        # self.reducers["dd"].direct_beam.m_spec
+        # self.reducers["dd"].direct_beam.m_spec_sd
+        # THIS IS WHERE THE MAGIC HAPPENS
+        self._efficiency_correction(poleff=poleff)
+
+        # once the wavelength spectra have been corrected/overwritten then the
+        # reflectivities need to be recalculated.
+        # this doesn't correct the offspecular
+        for sc in reduced_successfully:
+            reducer = self.reducers[sc]
+            # Add ycorr to reducer attributes
+            reducer.y_corr, reducer.y_corr_err = EP.EPdiv(
+                reducer.reflected_beam.m_spec_polcorr,
+                reducer.reflected_beam.m_spec_sd,
+                reducer.direct_beam.m_spec_polcorr,
+                reducer.direct_beam.m_spec_sd,
+            )
+            # now write out the corrected reflectivity files
+            fnames = []
+            datasets = []
+            datafilename = reducer.reflected_beam.datafilename
+            datafilename = os.path.basename(datafilename.split(".nx.hdf")[0])
+
+            for i in range(np.size(reducer.y, 0)):
+                data_tup = reducer.data(scanpoint=i)
```

Comment:
This data_tup won't actually hold corrected reflectivities. This is because `reducer.data` looks at `self.y, self.y_err` created as:

```
        reduction["x"] = self.x = xdata
        reduction["x_err"] = self.x_err = xdata_sd
        reduction["y"] = self.y = ydata / scale
        reduction["y_err"] = self.y_err = ydata_sd / scale
```

You've recalculated the reflectivity as `reducer.y_corr, y_corr_err`. This code should be something along the lines of:

```
data = reducer.data(scanpoint=i)
data_tup = (data.x, reducer.y_corr[i], reducer.y_corr_err[i], data[-1])
```

Reply:
This is now updated